runtime/caml/weak.h

```diff
@@ -176,7 +176,7 @@ Caml_inline void caml_ephe_clean_partial (value v,
     child = Field (v, i);
   ephemeron_again:
     if (child != caml_ephe_none
-        && Is_block (child) && Is_in_heap_or_young (child)){
+        && Is_block (child) && Is_in_value_area (child)){
```

Comment:
Contrarily to what is explained in the commit comment, this is also used in the case the block is white. Fortunately, this is fine, since an out-of-heap pointer should point to a black block.

However:
- Could you please update the commit comment (and remove the "to be merged", which does not seem to be relevant)
- This makes the property that out-of-heap pointers should be black more important than I have read somewhere else recently (the argument was that the GC would mark it as black during its first pass if it is left white). Is this clear in the docs (I have not checked)?

Reply:
I pushed a more detailed comment.  

As to out-of-heap pointers having black headers: yes, this turns out to be a crucial invariant, here and elsewhere, and not just an optimization to prevent useless marking.  The "detect naked pointers" variant of OCaml 4.10 that was recently released enforces this invariant.

As to documentation: currently, the "interfacing with C" chapter of the reference manual says that all out-of-heap pointers must be encapsulated (boxing or integer-like tagging), so it is not the place to document the black-header trick.